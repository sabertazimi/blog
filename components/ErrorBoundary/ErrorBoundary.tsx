import Paragraph from '@components/Paragraph'
import Result from '@components/Result'
import { Text } from '@components/Texts'
import type { ErrorInfo } from 'react'
import { Component } from 'react'

interface Props {
  children: JSX.Element
}

interface State {
  hasError: boolean
  error: Error
  errorInfo: ErrorInfo
}

class ErrorBoundary extends Component<Props, State> {
  readonly state: State = {
    hasError: false,
    error: Error('Error'),
    errorInfo: { componentStack: '' },
  }

  public componentDidCatch(error: Error, errorInfo: ErrorInfo): void {
    this.setState({
      hasError: true,
      error,
      errorInfo,
    })
  }

  public render() {
    const { hasError, error } = this.state
    const { children } = this.props
    // eslint-disable-next-line node/prefer-global/process -- SWC/Webpack supports process.env.NODE_ENV.
    const isDevelopment = process.env.NODE_ENV === 'development'

    if (hasError && isDevelopment) {
      return (
        <Result status="error" title="Some Error Happened">
          <div role="alert">
            <Paragraph>
              <Text className="text-xl font-extrabold">
                Please check the following information:
              </Text>
            </Paragraph>
            <Paragraph>
              <Text>{`${error.message}.`}</Text>
            </Paragraph>
            <Paragraph>
              <Text>Please check terminal output for further details.</Text>
            </Paragraph>
            <Paragraph>
              <Text>Please reload this page after changing code.</Text>
            </Paragraph>
          </div>
        </Result>
      )
    }

    return children
  }
}

export default ErrorBoundary
