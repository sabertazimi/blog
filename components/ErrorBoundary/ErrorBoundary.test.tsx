import { render, screen } from '@utils'
import { axe } from 'jest-axe'
import type { JSXElementConstructor, ReactNode } from 'react'
import ErrorBoundary from './ErrorBoundary'

describe('ErrorBoundary', () => {
  const ComponentWithError = ({ shouldThrow }: { shouldThrow?: boolean }) => {
    if (shouldThrow)
      throw new Error('ComponentWithError')
    else
      return <div>App</div>
  }
  const ENV = { ...process.env }
  let mockConsoleError: jest.SpyInstance

  beforeEach(() => {
    process.env = { ...ENV, NODE_ENV: 'development' }
    mockConsoleError = jest
      .spyOn(console, 'error')
      .mockImplementation(jest.fn())
  })

  afterEach(() => {
    process.env = ENV
    mockConsoleError.mockRestore()
  })

  it('should render children correctly (snapshot)', () => {
    const { container } = render(
      <ErrorBoundary>
        <ComponentWithError />
      </ErrorBoundary>,
    )

    expect(container).toMatchSnapshot()
  })

  it('should render alert message correctly (snapshot)', () => {
    const { container } = render(
      <ErrorBoundary>
        <ComponentWithError shouldThrow />
      </ErrorBoundary>,
    )

    expect(container).toMatchSnapshot()
  })

  it('should render children accessibility guidelines (AXE)', async () => {
    const { container } = render(
      <ErrorBoundary>
        <ComponentWithError />
      </ErrorBoundary>,
    )

    const a11y = await axe(container)

    expect(a11y).toHaveNoViolations()
  })

  it('should render alert message accessibility guidelines (AXE)', async () => {
    const { container } = render(
      <ErrorBoundary>
        <ComponentWithError shouldThrow />
      </ErrorBoundary>,
    )

    const a11y = await axe(container)

    expect(a11y).toHaveNoViolations()
  })

  it('should render alert message when error happened', () => {
    const { rerender } = render(<ComponentWithError />, {
      wrapper: ErrorBoundary as JSXElementConstructor<{ children: ReactNode }>,
    })

    expect(screen.queryByRole('alert')).not.toBeInTheDocument()

    rerender(<ComponentWithError shouldThrow />)

    expect(screen.getByRole('alert')).toBeInTheDocument()
    expect(mockConsoleError).toHaveBeenCalledTimes(3)
  })
})
