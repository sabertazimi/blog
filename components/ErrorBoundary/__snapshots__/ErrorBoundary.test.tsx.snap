// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary should render alert message correctly (snapshot) 1`] = `
<div>
  <div
    class="ant-result ant-result-error css-dev-only-do-not-override-j0nf2s"
  >
    <div
      class="ant-result-icon"
    >
      <span
        aria-label="close-circle"
        class="anticon anticon-close-circle"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="close-circle"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
          />
        </svg>
      </span>
    </div>
    <div
      class="ant-result-title"
    >
      Some Error Happened
    </div>
    <div
      class="ant-result-content"
    >
      <div
        role="alert"
      >
        <div
          class="ant-typography p dark:text-light css-dev-only-do-not-override-j0nf2s"
        >
          <span
            class="ant-typography text-xl font-extrabold dark:text-light css-dev-only-do-not-override-j0nf2s"
          >
            Please check the following information:
          </span>
        </div>
        <div
          class="ant-typography p dark:text-light css-dev-only-do-not-override-j0nf2s"
        >
          <span
            class="ant-typography dark:text-light css-dev-only-do-not-override-j0nf2s"
          >
            ComponentWithError.
          </span>
        </div>
        <div
          class="ant-typography p dark:text-light css-dev-only-do-not-override-j0nf2s"
        >
          <span
            class="ant-typography dark:text-light css-dev-only-do-not-override-j0nf2s"
          >
            Please check terminal output for further details.
          </span>
        </div>
        <div
          class="ant-typography p dark:text-light css-dev-only-do-not-override-j0nf2s"
        >
          <span
            class="ant-typography dark:text-light css-dev-only-do-not-override-j0nf2s"
          >
            Please reload this page after changing code.
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ErrorBoundary should render children correctly (snapshot) 1`] = `
<div>
  <div>
    App
  </div>
</div>
`;
